cocina_level2_prs.rb

```diff
@@ -0,0 +1,185 @@
+#!/usr/bin/env ruby
+# frozen_string_literal: true
+
+# To use this script, you must
+#  - install "hub" on your local machine:  https://hub.github.com/
+#  - have a github access token with scopes of "read:org" and "repo"
+#  (see https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token)
+#
+# Usage:
+# "GITHUB_TOKEN=abc123 ./cocina_level2_prs.rb
+
+# Cocina Level 2 updates are infrastructure apps that need to be updated after
+#   cocina-models gem new release
+#   and after dor-services-client, sdr-client gems new release
+#   and after dor-services-app, sdr-api updated (and api updated)
+require 'yaml'
+require 'bundler/setup'
+require 'open3'
+require 'fileutils'
+
+$LOAD_PATH.unshift 'lib'
+require 'error_emitting_executor'
+
+WORK_DIR = 'tmp/repos'
+BRANCH_NAME = 'cocina-level2-updates'
+GIT_MAIN_FETCH_REFS = '+refs/heads/main:refs/remotes/origin/main'
+GIT_BRANCH_FETCH_REFS = GIT_MAIN_FETCH_REFS.gsub('main', BRANCH_NAME).freeze
+COMMIT_DESCRIPTION = 'Update dependencies for cocina-models update'
+EXPECTED_PUSH_MESSAGE = "Branch '#{BRANCH_NAME}' set up to track remote branch '#{BRANCH_NAME}' from 'origin'"
+
+def repos_file
+  File.join 'infrastructure', 'projects.yml'
+end
+
+# @return [Array<String>] the git repos that need cocina level2 updates
+def repos
+  YAML.load_file(repos_file).fetch('projects')
+      .select { |project| project.fetch('cocina_level2', true) }
+      .map { |project| project.fetch('repo') }
+end
+
+def create_prs(repos)
+  repos.map do |repo|
+    pr_link = prepare_and_create_pr(repo)
+    p pr_link
+    { repo: repo, pr_link: pr_link }
+  end
+end
+
+def prepare_and_create_pr(repo)
+  clone_or_update_repo(repo)
+
+  within_cloned_repo_dir(repo) do
+    result = create_branch
+    result = update_gems unless result_failure?('create_branch', result)
+    result = local_git_commit unless result_failure?('update_gems', result)
+    result = git_push_origin unless result_failure?('local_git_commit', result)
+    create_pr if result.include?(EXPECTED_PUSH_MESSAGE)
+  end
+end
+
+# must be called as part of the block for within_cloned_repo_dir(repo)
+def create_branch
+  ErrorEmittingExecutor.execute("git checkout -B #{BRANCH_NAME}")
```

Comment:
This is trivial, but I'm curious about using `git checkout` up here to create/switch to a branch and then using `git switch` to switch to `main` below. If we wanted to be consistent we could also use `git switch -C` here. I don't think it's very important, though.

Reply:
very true.  I stuck with checkout at first because it covers old git versions.   Then I said "meh" when I did "switch".  Consistency is probably warranted.

Reply:
OK, I'll change it.